libcnb/src/build.rs

```diff
@@ -109,44 +110,57 @@ impl<B: Buildpack + ?Sized> BuildContext<B> {
         })
     }
 
+    /// Reads the [`LayerData`] for the layer with the given name.
+    ///
+    /// Returns `Ok(None)` is the layer doesn't exist.
```

Comment:
I'm not 💯 sure what you mean. My guess is that you want to remove the `Option` and make a non-existent layer an error variant of `ReadLayerError`? Because it is already is a `Result`.

In the end it's really just the decision if we consider a missing layer an error or a valid and expected outcome of this function. You can find arguments for both and I think there is no clear winner. Which solution is the "better" solution heavily depends on context.

For example, for `std::fs`, a missing file is an error which I think is reasonable as this mirrors what the operating systems do when accessing a missing file. The other extreme would be a collection API, where you wouldn't expect a `Result` for retrieving a value by key from a hash map.

In this case, I chose to signal a non-existing layer via an `Option`. However, for this function to operate, I have to parse data and do IO. As these operations are fallible, I need to bubble them up and wrap the result of the function as a `Result`. This make the final return type a `Result<Option<_>, _>`.

> Introducing a third would mean one fewer operation to unwrap and get the contents.

I think the ergonomic of that return type are alright. Some examples:

When you really care about all cases, layer existing, layer missing, specific errors and a catch-all for all other errors:

```rust
match result {
    Ok(None) => /* */,
    Ok(Some(layer)) => /* */,
    Err(SpecificError) => /* */,
    Err(_) => /* */
}
```
When you only care about the layer being there or not:

```rust
if let Some(layer) = read()? {
    /* */
} else {
    /* */
}
```

When you absolutely need the layer to exist and you have no way to recover (I don't think this will ever be the case for a real-world buildpack):

```rust
let layer = read()?.ok_or(BuildpackError::IReallyNeedThisLayer)?;
```

I'm not convinced that this needs changing, but happy to be convinced otherwise either by you or other folks :)

Reply:
> My guess is that you want to remove the Option and make a non-existent layer an error variant of ReadLayerError? Because it is already is a Result.

Yes, exactly. I was considering that out loud. I came to the conclusion that I think I actually like the `Result<Option>` 

> I'm not convinced that this needs changing, but happy to be convinced otherwise either by you or other folks :)

Good. I'm of the same persuasion. I left the comment to have it validated. We can resolve this.